File: 904-leaf-similar-trees/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        
        def root_leaf(root):
            result=[]
            if root :
                if  not root.left and not root.right:
                    return [root.val]
                result.extend(root_leaf(root.left))
                result.extend(root_leaf(root.right))
            return result
        
        root1_leaves = root_leaf(root1)
        root2_leaves = root_leaf(root2)

        return root1_leaves == root2_leaves

File: 904-leaf-similar-trees/leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void traverse(TreeNode root, ArrayList<Integer> res){
        if(root == null){
            return;
        }
        if(root.left == null && root.right == null){
            res.add(root.val);
        }
        traverse(root.left, res);
        traverse(root.right, res);
        return;
    }

    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        ArrayList<Integer> rl1 = new ArrayList<Integer>();
        ArrayList<Integer> rl2 = new ArrayList<Integer>();

        traverse(root1, rl1);
        traverse(root2, rl2);

        return rl1.equals(rl2);
    }
}